qadence/engines/jax/differentiable_expectation.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Tuple
+
+import jax.numpy as jnp
+from horqrux.utils import prepare_state
+from jax import Array, custom_vjp
+
+from qadence.backend import Backend as QuantumBackend
+from qadence.backend import ConvertedCircuit, ConvertedObservable
+from qadence.backends.utils import (
+    tensor_to_jnp,
+)
+from qadence.blocks.utils import uuid_to_eigen
+from qadence.measurements import Measurements
+from qadence.mitigations import Mitigations
+from qadence.noise import Noise
+from qadence.types import Endianness, Engine, ParamDictType
+
+
+def compute_gap(eigen_vals: Array) -> Array:
+    gaps = jnp.unique(jnp.abs(jnp.tril(eigen_vals - eigen_vals.reshape(-1, 1))))
+    # We have to filter out zeros
+    return gaps[gaps > 0]
+
+
+@dataclass
+class JaxDifferentiableExpectation:
+    """A handler for differentiating expectation estimation using various engines."""
+
+    backend: QuantumBackend
+    circuit: ConvertedCircuit
+    observable: list[ConvertedObservable] | ConvertedObservable
+    param_values: ParamDictType
+    state: Array | None = None
+    measurement: Measurements | None = None
+    noise: Noise | None = None
+    mitigation: Mitigations | None = None
+    endianness: Endianness = Endianness.BIG
+    engine: Engine = Engine.JAX
+
+    def psr(self) -> Any:
+        observable = self.observable[0]
+        if self.state is None:
+            self.state = prepare_state(
+                self.circuit.abstract.n_qubits, "0" * self.circuit.abstract.n_qubits
+            )
+
+        def _expectation_fn(
+            state: Array, values: dict, psr_params: dict, uuid_to_eigen: dict
+        ) -> Array:
+            wf = self.circuit.native.forward(state, values)
+            return observable.native.forward(wf, values)
+
+        @custom_vjp
+        def _expectation(
+            state: Array, values: dict, psr_params: dict, uuid_to_eigen: dict
+        ) -> Array:
```

Comment:
Is this intended to be used for finite shots too? If so, then how would `jax` get the relevant `PRNGKey` to introduce noise to the shots?

Reply:
i am planning to add it but it wont be in the first version of the JAX engine in qadence